### ISSUE TITLE
Fields can be undefined but return type of getters can't

### ISSUE THREAD

**Issue by renkei** [thesayyn/protoc-gen-ts#80]
What is the idea behind the current implementation to generate code from proto files where the return type of the field getters is always cast to the data type of the field?

The cast in the generated typescript code is realized with an `as` expression, which is dangerous because the Typescript compiler is happy (maybe) but it could happen that the resulting data type has nothing to do with the real data type available at runtime in JavaScript.

This is exactly the case if a field of a proto message B is of type proto message A.

```proto
message A
{
    int32 id = 1;
}

message B
{
    A item = 1;
}
```

If I use the generated typescript code to create an instance of B like:

```typescript
const b = new B();
```

then the generated typescript code tells me that `b.item` is always of type `A` but in fact, `b.item` is `undefined` here. Can be easily checked by printing `b.item` to `console`. So, the correct typings for `b.item` would be `A | undefined`. Or am I wrong? Do I miss something here?

**Comment by renkei** [comment]
It's even worse. Let's say, I know that the Typescript Compiler can't help me here because of the wrong typings in the generated code but it's OK, because I will do the check manually in the code. Then ESLint tells me that I'm doing wrong:

```typescript
// Unnecessary conditional, the types have no overlap. (eslint@typescript-eslint/no-unnecessary-condition)
if (b.item === undefined) {
    ...
}

// Unnecessary conditional, value is always truthy. (eslint@typescript-eslint/no-unnecessary-condition)
if (b.item) {
    ...
}
```

**Comment by thesayyn** [comment]
I am not sure if that is the correct approach here as we tried that one before.  The best would be optional getters which is not supported by typescript yet.

See #20, #33   

also microsoft/TypeScript#14417

**Comment by thesayyn** [comment]
I have no problem whatsoever with the approach above. there are some typescript limitations that prevents us from doing this.

The problem with getter setter types is that they can’t have different types. hence, since `T | undefined` is different than `T` type-checker will yell at you.

**Comment by thesayyn** [comment]
https://www.typescriptlang.org/play?#code/MYGwhgzhAECyCmUwHN7QN4ChrQvALtPovgBQBuYIArvAFy74BOAlgHbLQA+01bAJvABm7ePwCUGHAF9s0VIWIQy4hstYcMs6UA

Also see: https://github.com/microsoft/TypeScript/issues/2521

**Comment by thesayyn** [comment]
it is going to take a while this gets stabilized. then we can revisit the options.

**Comment by renkei** [comment]
Ok, that's true, the type of getters and setters must be equal. If I look at my check list above then only point 7 (`Well-Known Types`) is affected by this limitation. So, the setter must also accept `undefined`. That's not really an issue as long as the getter returns the correct typings so that an user of your package can safely work with returned values.

The main reason for this ticket is point 2 (`Singluar Message Types`) because this is really an issue. Here, the typings do not reflect the returned types at runtime. So, a customer of your package has to find a workaround.

To make it short: Accept `undefined` for setters of `Well-Known Types`, then you have the situation that all 7 items from the check list above have the same types for getters and setters. So everything can solved with current TypeScript versions. And this ticket could be closed once you switch from `T` to `T | undefined` for getters and setters of `Scalar Message Types`. Sorry, I still don't see the need to wait for new TypeScript features. Maybe I'm wrong.

**Comment by thesayyn** [comment]
> Sorry, I still don't see the need to wait for new TypeScript features. Maybe I'm wrong.

You are right. This definitely should be fixed. Would be happy to see a PR if you would like to take it. Otherwise, I will try to take this, this week.

**Comment by thesayyn** [comment]
Though this is going to be a breaking change. so I have to find a way to land this without annoying everyone.

**Comment by thesayyn** [comment]
> 7. [Well-Known Types](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf), e.g. `Int32Value`
>    For unset singular scalar types you (as a receiver / consumer of a message) cannot detect if the sender / producer of the mesage has set a default value or left the field unset. With e.g. `Int32Value` instead of `int32` this can be detected, so setters should accept `T` (no support for unset, the user can just set a value or ignore the field) and getters should return `T | undefined`

with proto3 optionals fields also get a getter called `has_[name]`
